### ISSUE TITLE
GAN design

### ISSUE THREAD

**Issue by zchen0211** [PaddlePaddle/Paddle#4563]
GAN is an important concept to support in our refactoring. Based on the current support of our framework, it is better to implement GANs with following concepts:

1) Dependency analysis and sub-graph running;
2) Gradient with respect to some certain variables;
3) Multiple loss functions design.

Operators and functions required to build GAN:

Important building blocks | People in Charge | Required
-- | -- | --
convolution 2d (done) | Chengduo | Y
cudnn conv 2d (missing) | Chengduo | N
deconv 2d (missing) | Zhuoyuan, Zhihong | Y
cudnn deconv 2d (missing) | Zhuoyuan, Zhihong | N
batch norm (missing) | Zhuoyuan, Jiayi | Y
cudnn batch norm (missing) | Zhuoyuan, Jiayi | N
max-pooling (done) | ? | Y
fc (done) | ? | Y
softmax loss (done) | ? | Y
reshape op (done) | ? | Y
Dependency Engine (done) | Jiayi | Y *
Python API (done) | Longfei, Jiayi | Y *
Executor (done) | Tony | Y *
Multi optimizer (woking) | Longfei | Y *
Optimizer with any para | ? | Y *
Concat op (done) | ? | N (Cond)
Repmat op (done)

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
